lib/elixir/lib/kernel.ex

```diff
@@ -2198,6 +2198,54 @@ defmodule Kernel do
     :erlang.error(ArgumentError.exception(error_message))
   end
 
+  @doc """
+  Returns true if `term` is a structure; otherwise returns `false`.
+
+  Allowed in guard tests.
+  """
+  @doc guard: true
+  defmacro is_struct(term) do
+    case __CALLER__.context do
+      nil ->
+        quote do
+          case unquote(term) do
+            %_{} -> true
+            _ -> false
+          end
+        end
+
+      :match ->
+        invalid_match!(:is_struct)
+
+      :guard ->
+        quote(do: is_atom(:erlang.map_get(unquote(term), :__struct__)))
```

Comment:
Sinec this function starts with the `is_prefix`, it should not fail. You should write a check like this:

```elixir
is_struct(map) or is_map(map)
```

And then pass it a map. Because you will call map_get on a map without the `__struct__`, the whole clause will fail but it shouldn't. The proper guard is:

    is_map(unquote(term)) and :erlang.is_map_key(:__struct__, unquote(term)) and is_atom(:erlang.map_get(:__struct__, unquote(term)))

Reply:
Failures in guards aren't errors, but just match failure. You can check this out via:

```elixir
def foo(list) when hd(list) == 1, do: :ok
def foo(other), do: :error

foo(10) == :error
```

Reply:
Added test for that and fixed it.